File: src/main/java/com/chw/java/spark/SparkWC.java
package com.chw.java.spark;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.Optional;
import org.apache.spark.api.java.function.VoidFunction;

import java.util.Arrays;

import scala.Tuple2;

public class SparkWC {
    public static void main(String[] args) {
        System.out.println("spark word count ...");
        SparkConf conf = new SparkConf();
        conf.setMaster("local").setAppName("test");

        JavaSparkContext sc = new JavaSparkContext(conf);

        final JavaPairRDD<String, Integer> nameRdd = sc.parallelizePairs(
                Arrays.asList(new Tuple2<String, Integer>("zhangsan", 10),
                        new Tuple2<String, Integer>("lisi", 20), new Tuple2<String, Integer>("maliu", 12)));
        final JavaPairRDD<String, Integer> socreRdd = sc.parallelizePairs(
                Arrays.asList(new Tuple2<String, Integer>("zhangsan", 100),
                        new Tuple2<String, Integer>("lisi", 200), new Tuple2<String, Integer>("wangwu", 300)));

//        final JavaPairRDD<String, Tuple2<Integer, Integer>> joinRdd = nameRdd.join(socreRdd);
//        joinRdd.foreach(new VoidFunction<Tuple2<String, Tuple2<Integer, Integer>>>() {
//            @Override
//            public void call(Tuple2<String, Tuple2<Integer, Integer>> stringTuple2Tuple2) throws Exception {
//                System.out.println(stringTuple2Tuple2);
//            }
//        });

        final JavaPairRDD<String, Tuple2<Integer, Optional<Integer>>> lefJoinRdd = nameRdd.leftOuterJoin(socreRdd);
        lefJoinRdd.foreach(new VoidFunction<Tuple2<String, Tuple2<Integer, Optional<Integer>>>>() {
            @Override
            public void call(Tuple2<String, Tuple2<Integer, Optional<Integer>>> stringTuple2Tuple2) throws Exception {
                System.out.println(stringTuple2Tuple2);

                final Optional<Integer> integerOptional = stringTuple2Tuple2._2._2;

                if (integerOptional.isPresent()) {
                    System.out.println(integerOptional.get());
                } else {
                    System.out.println("空值");
                }
            }
        });
    }

    public String getName() {
        return SparkWC.class.getSimpleName();
    }
}

File: src/main/scala/com/chw/scala/spark/SparkWC.scala
package com.chw.scala.spark

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

/**
 * spark版本的词频统计
 */
object SparkWC {
  def main(args: Array[String]): Unit = {
    println("spark word count ...")

    val conf = new SparkConf()
      .setAppName("word_count")
      .setMaster("local[2]")
//      .setMaster("spark://127.0.0.1:7077")
    val sc = new SparkContext(conf)
    sc.setLogLevel("error")

    val txtFile = "input.txt"
    val txtData: RDD[String] = sc.textFile(txtFile)

//    val txtData=sc.parallelize(Array("one","two","three","one"))

    txtData.cache()

    val lineCount = txtData.count()
    println(s"lineCount=$lineCount")

    val wcData = txtData.flatMap(_.split(" "))
      .map((_, 1))
      .reduceByKey(_ + _)

    wcData.collect().foreach(println)

    sc.stop
  }
}
